First deliverable: RxJava, Singleton, MVVM Pattern and RecyclerView Pagination

ponchOlvera opened the pull request:
Implemented MVVM pattern.
Implemented a Singleton pattern in the Retrofit instance to prevent multiple instances.
Use RxJava to return an Observable when executing a HTTP request.
Implemented a simple pagination in the RecyclerView and populate Marvel characters in the list.

Commit:
fixed dependencies and implemented RxJava with Retrofit calls

ponchOlvera committed:
Added fragment using viewbinding

ponchOlvera committed:
FirstDeliverable: V-1.0.0

+ Added MVVM pattern
+ Created the recyclerview and adapter to include items from the marvel API service.
+ Implemented simple pagination in the recyclerview, requesting new characters and add them to the recycler.

ponchOlvera committed:
FirstDeliverable: V-1.0.1

+ Added observable retrofit request to be used in the next deliverables
+ Added missing loading state change when an error is observed by the view.

ponchOlvera committed:
FirstDeliverable: Version 1.0.2

+ Added internet connection validation.

ponchOlvera committed:
FirstDeliverable: V-1.0.3

+ Added retry button to be displayed when no connection is available.
+ Modified RxJava observable pattern in retrofit to use Single observable which will listen only to one data result and then the subscription will end.

pestrada commented on app/build.gradle:
```diff
@@ -30,20 +30,38 @@ android {
+
+  //Glide
+  implementation 'com.github.bumptech.glide:glide:4.12.0'
+  annotationProcessor 'com.github.bumptech.glide:compiler:4.12.0'
```
The suggestion here is to use variables for setting any library dependencies. Similar to what you did with `paging_version`

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/HeroesServices.kt:
```diff
@@ -11,5 +15,22 @@ interface HeroesServices {
+
+    @GET("characters/{characterId}/comics")
+    fun comicsByCharacter(
+        @Query("ts") ts: String,
```
Suggestion -> Let's use more descriptive names for variables. Example instead of `ts` use `timeStamp` or whatever the real name is.
```
@Query("ts") timeStamp: String,
```

ponchOlvera replied:
Done

pestrada commented on app/src/main/java/com/wizeline/heroes/NetworkClient.kt:
```diff
@@ -1,14 +1,26 @@
 
     fun getServices(): HeroesServices {
-        return Retrofit.Builder().apply {
+        return retrofitInstance!!.create(HeroesServices::class.java)
```
We should avoid the operator `!!` because it can throw a runtime exception.
```suggestion
       if (retrofitInstance == null) createRetrofitInstance()
       return retrofitInstance.create(HeroesServices::class.java)
```

ponchOlvera replied:
Done. created a separate method that creates a new Retrofit instance if the global instance is null and return it. Therefore, it is safe to return the nullable instance as ```retrofitInstance!!```

pestrada commented on app/src/main/java/com/wizeline/heroes/NetworkClient.kt:
```diff
@@ -1,14 +1,26 @@
+
+    private fun createRetrofitInstance(){
+        retrofitInstance = Retrofit.Builder().apply {
             baseUrl("https://gateway.marvel.com/v1/public/")
```
In general, we can use a single source of truth for defining API endpoints. Something like:
```
class Endpoint {
  const val BASE_URL = "https://gateway.marvel.com/v1/public/"
  consts val COMICS_URL = "characters/{characterId}/comics"
  // etc.
}
```

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/models/Character.kt:
```diff
@@ -0,0 +1,26 @@
+    @SerializedName("modified") val modified: String,
+    @SerializedName("comics") val comics: ComicList,
+    @SerializedName("thumbnail") val thumbnail: Thumbnail,
+    @SerializedName("resourceURI") val resourceURI: String
```
When the serialized name is exactly the same as the variable name. You can omit the serialized name.
`val name: String,`

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/models/ComicData.kt:
```diff
@@ -0,0 +1,9 @@
+
+data class ComicData(
+    @SerializedName("code") val code: Int
+) {
```
Let's omit the curly braces if not needed.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt:
```diff
@@ -0,0 +1,63 @@
+import androidx.fragment.app.Fragment
+import com.wizeline.heroes.R
+
+object ExtensionFunctions {
```
Maybe rename to `FragmentExtensions.kt` in order to convey their intent.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt:
```diff
@@ -0,0 +1,63 @@
+            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val networkCapabilities = connectivityManager.activeNetwork ?: return false
+            val actNw =
```
let's rename to `activeNetworkCapabilities` for a clearer name.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/abstract/PaginationScrollListener.kt:
```diff
@@ -0,0 +1,26 @@
+        if (!isLoading() && !isLastPage()) {
+            if (visibleItemCount + firstVisibleItemPosition >= totalItemCount && firstVisibleItemPosition >= 0) {
+                loadMoreItems()
+            }//                    && totalItemCount >= ClothesFragment.itemsCount
```
let's remove commented code.

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,66 @@
+) :
+    RecyclerView.Adapter<CharacterRecyclerViewAdapter.ViewHolder>() {
+    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val imgCharacter: ImageView = view.findViewById(R.id.imgCharacter)
```
Can we investigate a bit more why ViewBinding was not working here? For the next deliverable.

ponchOlvera replied:
Change will be implemented on Second Deliverable.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,66 @@
+            .load(imagePath)
+            .centerCrop()
+            .placeholder(R.drawable.ic_superhero_placeholder)
+            .into(holder.imgCharacter)
```
suggestion, move this code to inside ViewHolder on a function:
```
ViewHolder {
   fun bind(character: Character) {
     holder.tvCharacterName.text = character.name
        holder.tvCharacterDescription.text = description
        // etc
   }
}
```

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,66 @@
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val character = list.get(position)
+        val description =
+            if (character.description.isEmpty()) context.getString(R.string.description_not_available) else character.description
```
It's a good idea to move this kind of presentation logic to a separate class, in order to let the ViewHolder just render things. Some alternatives can be:
1.- Using a Presentation class:
```
class CharacterPresentation(private val character: Character) {

    fun formattedDescription(): Int {
        return if (character.description.isEmpty()) R.string.description_not_available else R.string.empty_string
    }
}
```
Then in ViewHolder:
```
holder.tvCharacterDescription.text = characterPresentation.formattedDescription()
```

or

2.- Using a Mapper class to transform the Domain Entity model to a UI Layer model.
```
class CharacterMapper(private val character: Character) {
   fun toCharacterItem() : CharacterItem {
      return CharacterItem(
          name = character.name,
          description = getDescription()
      )
   }
   
   private fun getDescription() : Int {
       if (character.description.isEmpty()) context.getString(R.string.description_not_available) else character.description
   }
}
```
Then in ViewHolder:
```
holder.tvCharacterDescription.text = characterItem.getDescription()
```

Any of these options will also allow you to increase code coverage by adding Unit Tests that verify the appropriate Description is used when it's empty or not.

ponchOlvera replied:
Done. I sued the Mapper method.

pestrada commented on app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt:
```diff
@@ -0,0 +1,132 @@
+            mMarvelViewModel.getCharacters(mRetrofitRepository, mOffset)
+        } else{
+            changeLoadingState(false)
+            Toast.makeText(context, "No internet connection, please connect to a internet Network", Toast.LENGTH_LONG).show()
```
Message should be a string XML resource instead for allowing reuse and language translations (if necessary)

ponchOlvera replied:
Done.

pestrada commented on app/src/main/java/com/wizeline/heroes/viewmodels/MarvelViewModel.kt:
```diff
@@ -0,0 +1,72 @@
+        compositeDisposable.add(repository.getCharacters(ts, apikey, hash, offset).observeOn(AndroidSchedulers.mainThread())
+            .subscribeOn(Schedulers.io())
+            .subscribe({
+                val currentViewState = marvelViewState.value!!
+                val marvelCharacterList = ArrayList<Character>()
+                it.data.results.forEach {
+                    marvelCharacterList.add(it)
+                }
+                currentViewState.characterList = marvelCharacterList
+                currentViewState.isListUpdated = true
+                _marvelViewState.value = currentViewState
```
I think most of this boilerplate updates can be replaced by creating an instance directly and remove any `currentViewState` variable. Since `marvelViewState` should be always the current state to render, there's no need for an additional variable. For example we could do it like this:
```
_marvelViewState.value = MarvelViewState(
    characterList = it.data.results,
    isListUpdated = true
)
```
Same with other references to update `marvelViewState.value`. We just pass around the required data through the instance constructor. Also, that way you will avoid the use of `!!` operator.

ponchOlvera replied:
Done.

pestrada reviewed (changes requested):
Left a few comments. Code meets requirements for the First Deliverable 🎊 Good job.
Let's just incorporate the suggestions before merging 👍

ponchOlvera committed:
FirstDeliverable: V-1.0.4

+ Added code review suggestions.
+ Created a CharacterMapper class to format the Character data from the data source and transform it to be shown in the UI.
+ Removed unused imports.
+ Added gradle variables for dependencies versions.

pestrada reviewed (approved):
Good job 👍
